File: Common/Utf8Encoding.vb
Public Class Utf8Encoding

    Public Shared Function Encode(ByVal CharsArray As Char(), ByVal CharsIndex As Integer, ByVal CharsLenght As Integer,
                                  ByVal BytesArray As Byte(), ByVal BytesIndex As Integer, ByVal BytesLength As Integer,
                                  ByVal BeginningOfFile As Boolean,
                                  ByRef CharsRead As Integer, ByRef BytesWritten As Integer) As Boolean
        CharsLenght += CharsIndex
        BytesLength += BytesIndex

        Dim CharsInitialIndex = CharsIndex
        Dim BytesInitialIndex = BytesIndex
        Dim T = New Byte(5) {}
        If BeginningOfFile Then
            CharsIndex -= 1
        End If
        For CharsIndex = CharsIndex To CharsLenght - 1
            If BytesIndex = BytesLength Then
                Exit For
            End If

            Dim Ch As Integer
            If BeginningOfFile And CharsIndex = CharsInitialIndex - 1 Then
                Ch = &HFEFF
            Else
                Ch = Strings.AscW(CharsArray(CharsIndex))
            End If

            If Ch < 128 Then
                BytesArray(BytesIndex) = CByte(Ch)
                BytesIndex += 1
                Continue For
            End If

            Dim I = 0

            Do
                T(I) = CByte((2 << 6) Or (Ch And ((1 << 6) - 1)))
                Ch >>= 6
                I += 1
                ' If we are ready, we will have I + 1 bytes for the character.
                ' So, there will be I + 2 bits occupied by the preamble of the first byte.
                ' And 8 - I - 2 = 6 - I bits will be remaining.
            Loop Until Ch < (1 << (6 - I))

            ' We are having 6 - I bits remaining.
            ' So we have to make 7 - I zeros at the end of the byte.
            T(I) = CByte((255 Xor ((1 << (7 - I)) - 1)) Or Ch)
            I += 1

            If BytesIndex + I >= BytesLength Then
                Exit For
            End If

            For I = I - 1 To 0 Step -1
                BytesArray(BytesIndex) = T(I)
                BytesIndex += 1
            Next
        Next
        ' 12345678
        ' 76543210

        CharsRead = CharsIndex - CharsInitialIndex
        BytesWritten = BytesIndex - BytesInitialIndex
        Return True
    End Function

    Public Shared Function Decode(ByVal BytesArray As Byte(), ByVal BytesIndex As Integer, ByVal BytesLength As Integer,
                                  ByVal CharsArray As Char(), ByVal CharsIndex As Integer, ByVal CharsLength As Integer,
                                  ByVal BeginningOfFile As Boolean,
                                  ByRef BytesRead As Integer, ByRef CharsWritten As Integer) As Boolean
        CharsLength += CharsIndex
        BytesLength += BytesIndex

        Dim CharsInitialIndex = CharsIndex
        Dim BytesInitialIndex = BytesIndex

        For BytesIndex = BytesIndex To BytesLength - 1
            If CharsIndex = CharsLength Then
                Exit For
            End If

            Dim B = BytesArray(BytesIndex)

            If (B >> 7) = 0 Then
                CharsArray(CharsIndex) = Strings.ChrW(B)
                CharsIndex += 1
                Continue For
            End If

            If ((B >> 6) And 1) = 0 Then
                Return False
            End If

            Dim I = 5
            Do Until ((B >> I) And 1) = 0
                I -= 1
                If I = 0 Then
                    Return False
                End If
            Loop

            I = 7 - I

            If BytesIndex + I >= BytesLength Then
                Exit For
            End If

            Dim Ch As Integer = (B << (I + 1)) >> (I + 1)

            For J As Integer = 1 To I - 1
                BytesIndex += 1
                B = BytesArray(BytesIndex)
                If (B >> 6) <> 2 Then
                    Return False
                End If
                Ch = (Ch << 6) Or ((B << 2) >> 2)
            Next

            ' Exclude Byte Order Mark (BOM).
            If Not (BeginningOfFile And CharsIndex = CharsInitialIndex And Ch = &HFEFF) Then
                CharsArray(CharsIndex) = Strings.ChrW(Ch)
                CharsIndex += 1
            End If
        Next
        ' 12345678
        ' 76543210

        BytesRead = BytesIndex - BytesInitialIndex
        CharsWritten = CharsIndex - CharsInitialIndex
        Return True
    End Function

End Class

File: Common/Utf8Encoding.cs
namespace Ks
{
    namespace Common
    {
        public class Utf8Encoding
        {
            public static bool Encode(char[] CharsArray, int CharsIndex, int CharsLenght, byte[] BytesArray, int BytesIndex, int BytesLength, bool BeginningOfFile, out int CharsRead, out int BytesWritten)
            {
                CharsLenght += CharsIndex;
                BytesLength += BytesIndex;

                var CharsInitialIndex = CharsIndex;
                var BytesInitialIndex = BytesIndex;
                var T = new byte[6];
                if (BeginningOfFile)
                    CharsIndex -= 1;
                for (; CharsIndex < CharsLenght; CharsIndex++)
                {
                    if (BytesIndex == BytesLength)
                        break;

                    var Ch = default(int);
                    if (BeginningOfFile & (CharsIndex == (CharsInitialIndex - 1)))
                        Ch = 0xFEFF;
                    else
                        Ch = CharsArray[CharsIndex];

                    if (Ch < 128)
                    {
                        BytesArray[BytesIndex] = (byte)Ch;
                        BytesIndex += 1;
                        continue;
                    }

                    var I = 0;

                    do
                    {
                        T[I] = (byte)((2 << 6) | (Ch & ((1 << 6) - 1)));
                        Ch >>= 6;
                        I += 1;
                    } while (Ch >= (1 << (6 - I)));

                    // We are having 6 - I bits remaining.
                    // So we have to make 7 - I zeros at the end of the byte.
                    T[I] = (byte)((255 ^ ((1 << (7 - I)) - 1)) | Ch);
                    I += 1;

                    if ((BytesIndex + I) >= BytesLength)
                        break;

                    for (I -= 1; I >= 0; I--)
                    {
                        BytesArray[BytesIndex] = T[I];
                        BytesIndex += 1;
                    }
                }
                // 12345678
                // 76543210

                CharsRead = CharsIndex - CharsInitialIndex;
                BytesWritten = BytesIndex - BytesInitialIndex;
                return true;
            }

            public static bool Decode(byte[] BytesArray, int BytesIndex, int BytesLength, char[] CharsArray, int CharsIndex, int CharsLength, bool BeginningOfFile, out int BytesRead, out int CharsWritten)
            {
                CharsLength += CharsIndex;
                BytesLength += BytesIndex;

                var CharsInitialIndex = CharsIndex;
                var BytesInitialIndex = BytesIndex;

                try
                {
                for (; BytesIndex < BytesLength; BytesIndex++)
                {
                    if (CharsIndex == CharsLength)
                        break;

                    var B = BytesArray[BytesIndex];

                    if ((B >> 7) == 0)
                    {
                        CharsArray[CharsIndex] = (char)B;
                        CharsIndex += 1;
                        continue;
                    }

                    if (((B >> 6) & 1) == 0)
                        return false;

                    var I = 5;
                    while (((B >> I) & 1) != 0)
                    {
                        I -= 1;
                        if (I == 0)
                            return false;
                    }

                    I = 7 - I;

                    if ((BytesIndex + I) >= BytesLength)
                        break;

                    var Ch = (B << (I + 1)) >> (I + 1);
                    for (var J = 1; J < I; J++)
                    {
                        BytesIndex += 1;
                        B = BytesArray[BytesIndex];
                        if ((B >> 6) != 2)
                            return false;
                        Ch = (Ch << 6) | ((B << 2) >> 2);
                    }

                    // Exclude Byte Order Mark (BOM).
                    if (!(BeginningOfFile & (CharsIndex == CharsInitialIndex) & (Ch == 0xFEFF)))
                    {
                        CharsArray[CharsIndex] = (char)Ch;
                        CharsIndex += 1;
                    }
                }
                // 12345678
                // 76543210
                }
                finally
                {
                BytesRead = BytesIndex - BytesInitialIndex;
                CharsWritten = CharsIndex - CharsInitialIndex;
                }

                return true;
            }
        }
    }
}
